subsys/bluetooth/services/gadgets.c

```diff
@@ -0,0 +1,531 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <bluetooth/conn.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/services/gadgets.h>
+
+#include <sys/byteorder.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(bt_gatt_gadgets, CONFIG_BT_GATT_GADGETS_LOG_LEVEL);
+
+#define GADGETS_TRANSACTION_TYPE_FIRST 0b00
+#define GADGETS_TRANSACTION_TYPE_CONT 0b01
+#define GADGETS_TRANSACTION_TYPE_LAST 0b10
+#define GADGETS_TRANSACTION_TYPE_CTRL 0b11
+
+#define GADGETS_ACK_ID_NO_ACK -1
+
+/** Base part of header found in all stream packets */
+struct gadgets_header_base {
+	uint8_t trxn_id : 4;
+	uint8_t stream_id : 4;
+	uint8_t ext : 1;
+	uint8_t ack : 1;
+	uint8_t type : 2;
+	uint8_t sequence_no : 4;
+};
+
+/** Header in first packet in a transaction */
+struct gadgets_header_first {
+	struct gadgets_header_base base;
+	uint8_t reserved;
+	uint8_t total_transaction_length[2];
```

Comment:
This is in effect a uint16_t, but using a uint8_t array because of endianness conversion. Adding a comment for this. Adding defines for magic value assignments.

Reply:
Possibly if it would be helpful we have set of the function to encoding data in requested endianness conversion like
`sys_get_be16` in `sys/byteorder.h`

Reply:
Thank you, I am using the conversion functions from sys/byteorder.h already. They operate on uint8_t[2] to hold the non-native endianness values, so to avoid casting I used such arrays in the structs directly.